### PR TITLE
[AIRFLOW-3057] add prev_*_date_success to template context

### PR DESCRIPTION
**Pull request by dstandish**
Make sure you have checked _all_ steps below.

### Jira

My PR addresses [[AIRFLOW-3057]](https://issues.apache.org/jira/browse/AIRFLOW-3057) - add macros that see only successful dags

### Description

Two new variables are added to template context: 
* `prev_execution_date_success`
* `prev_start_date_success`.

These return the exec / start date for the same task in prior _successful_ dag run.   

Lazy evaluation is employed so that the query to look up prev_ti is not executed unless a templated field references the variable in question.

To enable the above, I added `state` as a param to `DagRun.get_previous_dagrun` and `TaskInstance._get_previous_ti`, and added properties `previous_ti_success` and `previous_execution_date_success` and `previous_start_date_success` to `TaskInstance`.  

##### Success of dag run vs success of TI

These new context variables e.g. `previous_execution_date_success` will return the execution date of the same TI in the prior successful dag run.  Alternatively, I could have had them return the dates for the prior successful TI.  

I decided not to do that so that (1) no new indexes were required and (2) I could reuse the `get_previous_dagrun` method with only a small change.  The only time this would give a different result would be if a trigger rule allowed the dag run to be successful despite the task failing, or perhaps if a task was run manually without a dag run (not sure if this is possible?).  But these issues did not seem to justify the slightly more elaborate changes it would require.  In any case, I have made it clear in the docs that the execution date is derived from last successful dag run.

##### Naming
On naming, I initially was going to do `previous_successful_X`, but then I saw the ticket that already existed for this issue, and they had suggested `previous_X_success`, and I thought that was maybe better anyway, so that's what I went with.  

##### _get_previous_ti
Previously this will call `get_previous_scheduled_dagrun` if `dag.catchup` and `get_previous_dagrun` otherwise.  The `scheduled` version is more efficient -- it just looks for one row, instead of pulling many and ordering and taking one.
For my change, even if `dag.catchup` is True, I want to use `get_previous_dagrun` if either `state`  is given or `schedule_interval is None`.  
If you are asking for state (in this case success), then you are not looking for merely previous scheduled dag run -- you want the previous one with state.  
And if schedule interval is None (which could be true with dogs that are always externally triggered), then using `get_previous_scheduled_dagrun` won't find anything.
You could argue that if `dag.catchup` is True then we shouldn't be looking for previous successful or whatever.  But unless there is a very good reason, methods should do what they say they are going to do.

##### get_previous_dagrun
Here we just need to build filters dynamically because if state is not given then we don't want to add state to the filters.

### Tests

My PR adds 4 new tests:
* test_previous_ti
* test_previous_ti_success
* test_previous_execution_date_success
* test_previous_start_date_success

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

### Code Quality

- [x] Passes `flake8`

### CODE REVIEW

File: airflow/models/dagrun.py
Comment (dstandish) [inline airflow/models/dagrun.py]: We only filter on state if state is supplied.

**Review by pgagnon — COMMENTED** [review]
Looks good. As a general comment, could I suggest that you add type hints to your arguments, as the source code is in the process of being made mypy-compatible?

**Comment by dstandish** [comment]
@pgagnon thanks.  i updated with type hints.  in some cases this required some new imports e.g. Session, Union, Optional.  For Session I used sqlalchemy.orm.session.Session. 
States have type string because they are string class attributes not members of class State.

Some of the mypy tests required some weird changes so as not to upset it. E.g.:
* dag_run.py#L225 I added `assert isinstance(session, Session)  # mypy` so it would not try to access `query` on `None` object
* taskinstance.py#L570 Could not implicitly return `None`so I did ```return prev_ti and prev_ti.start_date``` instead of `if prev_ti: return prev_ti.start_date`
* had to add some more asserts in tests

File: airflow/models/dagrun.py
Comment (ashb) [inline airflow/models/dagrun.py]: This is an odd one - what does it do without it? We don't need this anywhere else.

Additionally can you keep `session` as the last kwarg please?

File: airflow/models/taskinstance.py
Comment (ashb) [inline airflow/models/taskinstance.py]: `session` last please.

File: airflow/models/taskinstance.py
Comment (ashb) [inline airflow/models/taskinstance.py]: The reasoning behind this `if` is hard to work out - could you expand on it with a comment saying why?

File: tests/models/test_taskinstance.py
Comment (ashb) [inline tests/models/test_taskinstance.py]:
```suggestion
```

No commented code committed please :)

File: tests/models/test_taskinstance.py
Comment (ashb) [inline tests/models/test_taskinstance.py]: These test would be better using `@parameterized` - check out test/jobs/test_scheduler_job.py for an example.

File: tests/models/test_taskinstance.py
Comment (ashb) [inline tests/models/test_taskinstance.py]: This feels like it shouldn't be needed, but I'm not familiar with mypy. What does it give without it?

File: tests/models/test_taskinstance.py
Comment (ashb) [inline tests/models/test_taskinstance.py]: Since each of these asserts are used once for each test I think the code would be better if this helper function returned the three ti's and the test cases did the asserts themselves.

File: airflow/models/taskinstance.py
Comment (ashb) [reply airflow/models/taskinstance.py]: Can you also confirm that the tests cover all three cases? (It's early and my ☕️  hasn't kicked in yet).

**Review by ashb — CHANGES_REQUESTED** [review]
Nice feature.

A little bit of a refactor of the tests please, but nice coverage!

File: airflow/models/dagrun.py
Comment (dstandish) [reply airflow/models/dagrun.py]: Here's an example of failing mypy test output: https://travis-ci.org/apache/airflow/jobs/542113316

In this case it tries to access the `query` method on a `None`.  I think it's because `None` is the default argument for `session`.  It did not fail until I added type hinting.

File: airflow/models/taskinstance.py
Comment (dstandish) [reply airflow/models/taskinstance.py]: Yeah I will add comment.  I don't like it but it's only because for whatever reason current state is to use `get_previous_scheduled_dagrun` when `dag.catchup=True`.  `get_previous_scheduled_dagrun` is a little more efficient because it just gets one row instead of order by limit 1.   But when we are filtering by state we don't care about schedule and same when there _is_ no schedule.

File: tests/models/test_taskinstance.py
Comment (dstandish) [reply tests/models/test_taskinstance.py]: Similar to above... because previous_ti can return `None` or `TI`, mypy apparently assumes that `None` should have `execution_date` attribute, since it is called in the next line.  It doesn't, so it failed.  The assert apparently tells mypy "in this situation, `previous_ti` will not be `None` so don't worry bout it".

File: tests/models/test_taskinstance.py
Comment (dstandish) [reply tests/models/test_taskinstance.py]: Yeah I thought about doing it that way, but it also uses the dates `d1`, `d2`, `d3`.  I thought about making a class and class attributes so that I would not have to return so many objects but I guess I can just throw them in a dictionary and call it a day.

Alternatively, could make a separate test class for these prev dates / TIs tests, make this helper the setup method, store the useful objects as attributes, where they can be used by these individual tests.  LMK your thoughts.  

But yeah either way I should make this better.

File: airflow/models/taskinstance.py
Comment (dstandish) [reply airflow/models/taskinstance.py]: Re coverage yes these cases are covered that's what this part is doing:
```
        schedule_interval_values = ('0 0 * * *', None, datetime.timedelta(days=1))
        catchup_values = (True, False)

        for tuple_ in product(schedule_interval_values, catchup_values):
            self._test_previous_dates(*tuple_, object_='previous_ti')
```
It's taking all combinations of the two lists and running the relevant tests for each combination.

**Comment by dstandish** [comment]
Thanks a bunch.  Responded to your comments.  Will work on the changes tomorrow.

File: tests/models/test_taskinstance.py
Comment (dstandish) [reply tests/models/test_taskinstance.py]: To be clear though each test calls the helper 6 times -- 3 schedule types * 2 values for catchup.

File: airflow/models/dagrun.py
Comment (ashb) [reply airflow/models/dagrun.py]: Pinging @jmcarp as our resident mypy "expert" :) Does this sound like a sensible change?

Comment (ashb) [reply airflow/models/dagrun.py]: Could you test something @dstandish: what happens if you remove the `=None` on the session and this assert? because of the decorator it will never be passed with None in anyway.

File: tests/models/test_taskinstance.py
Comment (ashb) [reply tests/models/test_taskinstance.py]: Just to check: does adding a `self.assertIsNotNone(t3.previous_ti)` (or `self.assertIsInstance(t3.previous_ti, TI)` keep mypy happy? (if it does those asserts give better diagnostics when they fail)

File: airflow/models/dagrun.py
Comment (pgagnon) [reply airflow/models/dagrun.py]: @dstandish By default mypy treats arguments which default to `None` as implicit `Optional[T]`. IMO this is a good thing because it adds a minimal degree of null-safety to typed Python, but it creates these confusing situations.

If as @ashb asserts the decorator will _always_ pass a `Session` object I would just remove the `isinstance` call and `None` default value to the `session` argument.

As a side note, there are situations where you positively need an argument with defaults to `None`, but where you know for a fact that in practice this will never happen (i.e. some usages of `functools.partial`) In these cases you could use a mypy `cast` here instead of `isinstance`. The difference being that `cast` is a dud at runtime.

Comment (ashb) [reply airflow/models/dagrun.py]: God to know @pgagnon

Comment (dstandish) [reply airflow/models/dagrun.py]: ok so @ashb shall i just change it to `cast` or shall I remove default argument of `None`? if i remove `None`, i will not be able to leave `session` the last argument.

File: tests/models/test_taskinstance.py
Comment (dstandish) [reply tests/models/test_taskinstance.py]: I did try this -- mypy doesn't like it.  i will change to `cast`.

File: airflow/models/dagrun.py
Comment (ashb) [reply airflow/models/dagrun.py]: Good point. Use `cast` then.

Comment (dstandish) [reply airflow/models/dagrun.py]: OK so it turns out "using `cast`" means, for example, adding this line instead of the assert: `session = cast(Session, session)`.  Also pretty ugly!   Alas, it seems like what we have to do.

Comment (pgagnon) [reply airflow/models/dagrun.py]: @ashb I think it might be preferable to remove the `None` argument instead of using `cast`, because it makes the function's signature more accurate.

edit: I'll defer to your final design decision, of course. ;-)

Comment (ashb) [reply airflow/models/dagrun.py]: @pgagnon Yes, but then it can't be after other optional arguments, and the entire rest of the codebase where we use `@provide_session` has `session=None` as the last kwarg in functions.

Comment (pgagnon) [reply airflow/models/dagrun.py]: @ashb Ah I see. That trumps my concern then.

Comment (ashb) [reply airflow/models/dagrun.py]: I think I'd rather we kept with convention on this PR and changed everything wholesale if we want to make a change.

Comment (dstandish) [reply airflow/models/dagrun.py]: Did we decide to type everything everywhere?  I am perfectly happy to skip typing in this PR if it is raising problems we're not keen to deal with now, if we want to later fix in a universal way.
Of course, I can see a benefit to adding typing whenever you touch something, if the decision is to add typing everywhere.

Comment (ashb) [reply airflow/models/dagrun.py]: Add typing as we edit bits, but don't change the pattern we use for `session=None`

File: airflow/models/taskinstance.py
Comment (dstandish) [reply airflow/models/taskinstance.py]: > The reasoning behind this `if` is hard to work out - could you expand on it with a comment saying why?

This is what I ended up with:
```
# We always ignore schedule in dagrun lookup when `state` is given or `schedule_interval is None`.
# For legacy reasons, when `catchup=True`, we use `get_previous_scheduled_dagrun` unless `ignore_schedule`.
ignore_schedule = state is not None or dag.schedule_interval is None
if dag.catchup is True and not ignore_schedule:
    last_dagrun = dr.get_previous_scheduled_dagrun(session=session)
else:
    last_dagrun = dr.get_previous_dagrun(session=session, state=state)
```

With the `ignore_schedule` logic pulled out I think it is much easier to read.

**Discussion**

With status quo, if `catchup is True`, `previous_ti` is "last ti according to schedule, if one exists", but if `catchup is False`, previous_ti is "last ti absolutely, if one exists".  

When we do the spring cleaning of context params in 2.0, we should fix this and call things what they are, and not have behavior subtly change based on other params if it can be avoided.  

E.g. I would say `previous_ti` should _always_ give the absolute last ti without regard to schedule, and we should make a separate property `previous_scheduled_ti` for the last _scheduled_ ti.  Then these methods would behave the same no matter `catchup` or `schedule_interval`; one would call prev dag run, the other would call prev scheduled dag run.  I would be happy to separate these in this manner in a subsequent PR if you like this idea.

**Note on the `schedule is None` and `catchup=True` case**
In status quo, for this odd case `previous_ti` _always_ returns `None`, because it is using `get_previous_scheduled_dagrun` when there is no schedule to use.  I discovered this when adding tests for `previous_ti`.  This PR fixes that by using `get_previous_dagrun` in this case.

Please LMK thoughts!   In particular:
* Are these comments sufficient?
* Do you like the idea of splitting previous_ti (absolute) and previous_scheduled_ti (scheduled)?

Comment (dstandish) [reply airflow/models/taskinstance.py]: OK this is actually a bit more readable:
```
# We always ignore schedule in dagrun lookup when `state` is given or `schedule_interval is None`.
# For legacy reasons, when `catchup=True`, we use `get_previous_scheduled_dagrun` unless `ignore_schedule`.
ignore_schedule = state is not None or dag.schedule_interval is None
if dag.catchup is True and not ignore_schedule:
    last_dagrun = dr.get_previous_scheduled_dagrun(session=session)
else:
    last_dagrun = dr.get_previous_dagrun(session=session, state=state)
```

**Comment by dstandish** [comment]
OK -- I did a serious refactor of the tests, used `parameterized` etc -- looking much better now.

I think everything else should be good to go as well.  I have left thread  https://github.com/apache/airflow/pull/5372#discussion_r291314335 unresolved so you can review comment wording and I make a suggestion for further de-muddying the logic here in a future PR.

File: tests/models/test_taskinstance.py
Comment (ashb) [inline tests/models/test_taskinstance.py]: Could we do:

```suggestion
            setup_dict['ti3_s'].execution_date
```

then we wouldn't need to return the dates directly, just the task instances? (It took me longer than I like to admit to work out what the `ed` prefix meant!)

File: tests/models/test_taskinstance.py
Comment (ashb) [inline tests/models/test_taskinstance.py]: It obviously isn't a requirement as the tests are passing but
```suggestion
            ti = TI(task=task, dag_id=dag.id, execution_date=pendulum.parse(execution_date))
```

File: tests/models/test_taskinstance.py
Comment (ashb) [inline tests/models/test_taskinstance.py:1006]:
```suggestion
            dag.create_dagrun(
                dag_id=dag.id
```

File: tests/models/test_taskinstance.py
Comment (ashb) [inline tests/models/test_taskinstance.py:1059]: Overall these tests are much nicer! However there is a case missing - we don't check the behaviour when there is no successful previous TI at all. 

Using the `param` fn from Parameterized we could do something like this:

```python
    @parameterized.expand((
        ('cron/catchup', '0 0 * * * ', True),
        ('cron/no-catchup', '0 0 * * *', False),
        ('no-sched/catchup', None, True),
        ('no-sched/no-catchup', None, False),
        ('timedelta/catchup', datetime.timedelta(days=1), True),
        ('timedelta/no-catchup', datetime.timedelta(days=1), False),
        param('something_previous_failed', datetime.timedelta(days=1), False, all_previous_failed=True),
    ))
```



```suggestion
    def test_previous_ti_success(self, _, schedule_interval, catchup, all_previous_failed=False) -> None:
```

(and then use all_previous_failed in `_test_previous_dates_setup` and to change the assertions we run etc.)

**Review by ashb — CHANGES_REQUESTED** [review]
Looking better! I can now see that we've got a case missing in the tests.

File: airflow/models/taskinstance.py
Comment (ashb) [reply airflow/models/taskinstance.py]: > E.g. I would say `previous_ti` should always give the absolute last ti without regard to schedule, and we should make a separate property `previous_scheduled_ti` for the last scheduled ti.

That sounds sensible for 2.0, yes. Can you reply to the post on the mailing list about context params with this change/suggestion? (I'm planning to cherry pick this PR to 1.10.4 once we get it merged, so we can't change the behaviour yet)

Comments LGTM

File: tests/models/test_taskinstance.py
Comment (dstandish) [reply tests/models/test_taskinstance.py:1059]: So we already test that something previous failed.  This is what I get for not using verbose variable names. ;) 

The `number` in `ti{number}_{letter}` means sequence in time.  The `letter` means failed or successful.  The 3 test TIs are `ti1_s`, `ti2_f`, `ti3_s`.

Because `ti3_s` returns `ti1_s` _(skipping over the failed TI `ti2_f`)_, we verify that it skips failed TIs.

And because `ti1_s` returns `None`, we verify that when it finds no prior successful runs, `None` is the value returned.

If you still think we're missing a case LMK!

Comment (ashb) [reply tests/models/test_taskinstance.py:1059]: I think we are missing when _all_ previous TIs failed. (i.e. previous_ti_success would return None). The subtle difference between this case and `ti1_s` is that ti1_s didn't have any previous ones (successful or otherwise).

Comment (dstandish) [reply tests/models/test_taskinstance.py:1059]: OK so how about I add a `ti0_f`.  Then this case is covered by the first assert, which checks that `ti1_s.previous_ti_success` returns `None`.

File: tests/models/test_taskinstance.py
Comment (dstandish) [reply tests/models/test_taskinstance.py:1006]: create_dagrun does not take dag_id arg

File: tests/models/test_taskinstance.py
Comment (dstandish) [reply tests/models/test_taskinstance.py]: TI does not take dag_id arg it gets it from task

File: tests/models/test_taskinstance.py
Comment (dstandish) [reply tests/models/test_taskinstance.py]: > It took me longer than I like to admit to work out what the ed prefix meant!
I am sorry, I can make verbose if you think better.  I should have done `ed_1_s` in hindsight.

The reason I used the actual dates is to ensure that these comparisons were not trivially equal because of some default under the hood.  E.g., I figured,  maybe every task get's a default execution_date under certain circumstances.  Then they could pass tests but they are not actually the correct dates.  

If you think this is silly I can chop the dates. LMK!

File: tests/models/test_taskinstance.py
Comment (ashb) [reply tests/models/test_taskinstance.py:1006]: Good point :)

File: tests/models/test_taskinstance.py
Comment (ashb) [reply tests/models/test_taskinstance.py]: ashb: 0:
dstandish: 2

😁

File: tests/models/test_taskinstance.py
Comment (dstandish) [reply tests/models/test_taskinstance.py:1059]: OK -- I refactored the setup helper to allow you so set up arbitrary scenarios in a given test, doing away with the static constants.  

This is way better.  Thank you for shepherding me through this learning process and forcing me to do this better.  Never really written tests before!

Anyway, now a test looks like so:
```
        scenario = [State.SUCCESS, State.FAILED, State.SUCCESS]

        ti_list = self._test_previous_dates_setup(schedule_interval, catchup, scenario)

        self.assertIsNone(ti_list[0].previous_ti)

        self.assertEqual(
            ti_list[2].previous_ti.execution_date,
            ti_list[1].execution_date
        )

        self.assertNotEqual(
            ti_list[2].previous_ti.execution_date,
            ti_list[0].execution_date
        )
```

Each test can set up whatever scenario it needs, in order to test desired behavior.  The indexes of `ti_list` are the same as the `scenario` list.

So, returning to your concern here, for `previous_ti_success`, I used the FSFS scenario, and it tests that `ti_list[1].previous_ti_success` is `None`.

Please take a look when you have a sec and let me know if you'd like anything further.

File: tests/models/test_taskinstance.py
Comment (dstandish) [reply tests/models/test_taskinstance.py]: This issue is rendered obsolete by latest update.  And to guard against trivial equality, I added some `assertNotEqual` asserts.  Marking resolved.

**Comment by dstandish** [comment]
@ashb fyi, just in case you missed notification, I think this is good to go (regarding all prior issues raised).   I have left the last conversation unresolved just in you think there is anything further needed there.

File: tests/models/test_taskinstance.py
Comment (ashb) [inline tests/models/test_taskinstance.py:1061]: Nice - this makes the test super clear!

**Review by ashb — APPROVED** [review]

**Pull request merged**
